File: disk_distribution_mt.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <functional>
#include <iomanip>
#include <iostream>
#include <mutex>
#include <queue>
#include <random>
#include <thread>
#include <unordered_map>
#include <vector>

using namespace std;

// 定义tag数组
const vector<int> tags = {1968, 799,  2121, 1849, 944, 502,  1933, 1507,
                          494,  1540, 631,  1715, 434, 1107, 545,  2196};

// 定义分配方案的数据结构
struct Split {
  int disk_idx;
  int portion; // 使用整数表示百分比
  Split(int d, int p) : disk_idx(d), portion(p) {}
};

using Distribution = unordered_map<int, vector<Split>>;

// 用于线程同步的互斥锁和全局最优解
mutex mtx;
double global_best_variance = INFINITY;
Distribution global_best_distribution;

// 计算磁盘负载
vector<double> calculate_disk_loads(const Distribution &distribution) {
  vector<double> disk_loads(10, 0.0);
  for (const auto &[tag_idx, splits] : distribution) {
    double tag_size = tags[tag_idx];
    for (const auto &split : splits) {
      disk_loads[split.disk_idx] += tag_size * split.portion / 100.0;
    }
  }
  return disk_loads;
}

// 检查分配是否有效
bool is_valid_distribution(const Distribution &distribution) {
  vector<int> disk_tags(10, 0);

  // 检查每个磁盘的tag数量和每个tag的总比例
  for (const auto &[tag_idx, splits] : distribution) {
    int total_portion = 0;
    for (const auto &split : splits) {
      if (split.disk_idx >= 10)
        return false;
      disk_tags[split.disk_idx]++;
      total_portion += split.portion;
    }
    if (total_portion != 100)
      return false;
  }

  // 检查每个磁盘的tag数量是否在1-4之间
  for (int count : disk_tags) {
    if (count < 1 || count > 4)
      return false;
  }

  return true;
}

// 计算评分（考虑多个因素）
double calculate_score(const vector<double> &disk_loads) {
  double total_load = 0.0;
  double min_load = INFINITY;
  double max_load = 0.0;

  for (double load : disk_loads) {
    total_load += load;
    min_load = min(min_load, load);
    max_load = max(max_load, load);
  }

  double mean = total_load / 10;
  double variance = 0.0;
  double max_deviation = 0.0;

  for (double load : disk_loads) {
    double diff = load - mean;
    variance += diff * diff;
    max_deviation = max(max_deviation, abs(diff));
  }

  // 调整权重以更注重均衡性
  return variance / 10 + max_deviation * 3 + (max_load - min_load) * 2;
}

// 生成所有可能的分割方案
vector<vector<int>> generate_all_portions(int num_splits) {
  vector<vector<int>> result;
  vector<int> current;

  function<void(vector<int> &, int, int, int)> generate_recursive =
      [&](vector<int> &current, int remaining, int parts_left, int min_val) {
        if (parts_left == 1) {
          if (remaining >= min_val) {
            current.push_back(remaining);
            result.push_back(current);
            current.pop_back();
          }
          return;
        }

        // 使用1%的步长
        for (int i = min_val; i <= remaining - min_val * (parts_left - 1);
             i += 1) {
          current.push_back(i);
          generate_recursive(current, remaining - i, parts_left - 1, min_val);
          current.pop_back();
        }
      };

  generate_recursive(current, 100, num_splits, 1); // 最小分割为1%
  return result;
}

// 模拟退火算法的温度函数
double temperature(int iteration, int max_iterations) {
  // 使用更缓慢的冷却速度
  return pow(0.99, iteration * max_iterations / 1000000.0);
}

// 初始化一个基本的分配方案
Distribution initialize_distribution() {
  Distribution dist;
  vector<double> disk_loads(10, 0.0);
  vector<int> disk_tag_counts(10, 0);

  // 按大小排序tags
  vector<pair<int, int>> sorted_tags;
  for (int i = 0; i < tags.size(); ++i) {
    sorted_tags.emplace_back(tags[i], i);
  }
  sort(sorted_tags.rbegin(), sorted_tags.rend());

  // 贪心分配
  for (const auto &[size, tag_idx] : sorted_tags) {
    // 找到最适合的磁盘
    int best_disk = -1;
    double min_score = INFINITY;

    for (int disk = 0; disk < 10; ++disk) {
      if (disk_tag_counts[disk] >= 4)
        continue;

      // 尝试分配到这个磁盘
      disk_loads[disk] += size;
      double score = calculate_score(disk_loads);
      disk_loads[disk] -= size;

      if (score < min_score) {
        min_score = score;
        best_disk = disk;
      }
    }

    vector<Split> splits;
    splits.emplace_back(best_disk, 100);
    dist[tag_idx] = splits;

    disk_loads[best_disk] += size;
    disk_tag_counts[best_disk]++;
  }

  return dist;
}

// 尝试优化分配方案
void optimize_distribution(int thread_id, int num_threads) {
  random_device rd;
  mt19937 gen(rd());
  uniform_int_distribution<> dis(0, 9);
  uniform_int_distribution<> tag_dis(0, tags.size() - 1);

  Distribution current_dist = initialize_distribution();
  const int iterations_per_thread = 100000; // 大幅增加迭代次数

  // 预生成所有可能的分割方案
  vector<vector<vector<int>>> all_portions;
  for (int i = 1; i <= 4; ++i) {
    all_portions.push_back(generate_all_portions(i));
  }

  // 记录线程局部最优解
  double local_best_score = calculate_score(calculate_disk_loads(current_dist));
  Distribution local_best_dist = current_dist;
  int stagnation_count = 0;

  for (int iter = 0; iter < iterations_per_thread; ++iter) {
    double temp = temperature(iter, iterations_per_thread);
    Distribution new_dist = current_dist;

    // 随机选择1-4个tag进行优化
    int num_tags_to_optimize = (iter % 4) + 1;
    vector<int> selected_tags;
    for (int i = 0; i < num_tags_to_optimize; ++i) {
      int tag_idx;
      do {
        tag_idx = tag_dis(gen);
      } while (find(selected_tags.begin(), selected_tags.end(), tag_idx) !=
               selected_tags.end());
      selected_tags.push_back(tag_idx);
    }

    // 对选中的tag尝试不同的分割方案
    for (int tag_idx : selected_tags) {
      vector<Split> &splits = new_dist[tag_idx];

      // 随机选择分割数量
      int num_splits = (gen() % 4) + 1;
      const auto &portions = all_portions[num_splits - 1];

      if (portions.empty())
        continue;

      // 随机选择一个分割方案
      const auto &selected_portions = portions[gen() % portions.size()];

      splits.clear();
      vector<int> available_disks;
      for (int i = 0; i < 10; ++i) {
        available_disks.push_back(i);
      }
      shuffle(available_disks.begin(), available_disks.end(), gen);

      for (int i = 0; i < num_splits; ++i) {
        splits.emplace_back(available_disks[i], selected_portions[i]);
      }
    }

    if (is_valid_distribution(new_dist)) {
      vector<double> disk_loads = calculate_disk_loads(new_dist);
      double new_score = calculate_score(disk_loads);
      double current_score =
          calculate_score(calculate_disk_loads(current_dist));

      // 模拟退火决策
      double delta = new_score - current_score;
      if (delta < 0 ||
          (temp > 0 &&
           exp(-delta / temp) > uniform_real_distribution<>(0, 1)(gen))) {
        current_dist = new_dist;
        current_score = new_score;

        // 更新局部最优解
        if (new_score < local_best_score) {
          local_best_score = new_score;
          local_best_dist = new_dist;
          stagnation_count = 0;
        }

        // 更新全局最优解
        lock_guard<mutex> lock(mtx);
        if (new_score < global_best_variance) {
          global_best_variance = new_score;
          global_best_distribution = new_dist;
        }
      }
    }

    // 如果局部最优解长时间没有改善，重新初始化
    stagnation_count++;
    if (stagnation_count > 10000) {
      current_dist = initialize_distribution();
      stagnation_count = 0;
    }
  }
}

// 多线程优化
pair<Distribution, double> find_optimal_distribution_mt() {
  const int num_threads = thread::hardware_concurrency();
  vector<thread> threads;

  // 初始化全局最优解
  global_best_distribution = initialize_distribution();
  global_best_variance =
      calculate_score(calculate_disk_loads(global_best_distribution));

  // 启动多个线程进行优化
  for (int i = 0; i < num_threads; ++i) {
    threads.emplace_back(optimize_distribution, i, num_threads);
  }

  // 等待所有线程完成
  for (auto &t : threads) {
    t.join();
  }

  return {global_best_distribution, global_best_variance};
}

// 打印结果
void print_result(const Distribution &distribution) {
  if (distribution.empty()) {
    cout << "未找到有效解" << endl;
    return;
  }

  vector<double> disk_loads = calculate_disk_loads(distribution);
  vector<vector<pair<int, double>>> disk_tags(10);

  for (const auto &[tag_idx, splits] : distribution) {
    for (const auto &split : splits) {
      double portion = split.portion / 100.0;
      disk_tags[split.disk_idx].emplace_back(tag_idx, portion);
    }
  }

  // 计算统计信息
  double total_load = 0.0;
  double min_load = INFINITY;
  double max_load = 0.0;

  for (double load : disk_loads) {
    total_load += load;
    min_load = min(min_load, load);
    max_load = max(max_load, load);
  }

  double mean_load = total_load / 10;
  double variance = 0.0;
  for (double load : disk_loads) {
    variance += (load - mean_load) * (load - mean_load);
  }
  variance /= 10;

  cout << "\n负载统计：" << endl;
  cout << "平均负载: " << fixed << setprecision(2) << mean_load << endl;
  cout << "最小负载: " << min_load << endl;
  cout << "最大负载: " << max_load << endl;
  cout << "负载差异: " << max_load - min_load << endl;
  cout << "方差: " << variance << endl;
  cout << "优化目标值: " << global_best_variance << endl;

  cout << "\n最优分配方案：" << endl;
  for (int disk_idx = 0; disk_idx < 10; ++disk_idx) {
    cout << "\n磁盘 " << disk_idx << ":" << endl;
    cout << "总负载: " << disk_loads[disk_idx] << endl;
    cout << "Tags:" << endl;
    for (const auto &[tag_idx, portion] : disk_tags[disk_idx]) {
      cout << "  Tag " << tag_idx << ": " << portion << " * " << tags[tag_idx]
           << " = " << portion * tags[tag_idx] << endl;
    }
  }
}

int main() {
  cout << "使用 " << thread::hardware_concurrency() << " 个线程进行计算"
       << endl;

  auto start_time = chrono::high_resolution_clock::now();

  auto [best_distribution, best_variance] = find_optimal_distribution_mt();

  auto end_time = chrono::high_resolution_clock::now();
  auto duration =
      chrono::duration_cast<chrono::milliseconds>(end_time - start_time);

  print_result(best_distribution);

  cout << "\n程序运行时间: " << duration.count() << " 毫秒" << endl;

  return 0;
}

File: IO/model/DiskDistributionMT.java
package IO.model;

import java.text.DecimalFormat;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class DiskDistributionMT {

    // 定义 tag 数组
    private static int[] tags = { 1968, 799, 2121, 1849, 944, 502, 1933, 1507, 494, 1540, 631,
            1715, 434, 1107, 545, 2196 };
    // 分配方案: tag 索引 -> 若干拆分
    private static final Map<Integer, List<Split>> globalBestDistribution = new HashMap<>();
    private static final Lock mtx = new ReentrantLock();
    private static double globalBestVariance = Double.POSITIVE_INFINITY;

    // 定义分配方案的数据结构
    public static class Split {
        public int diskIdx;
        public int portion; // 拆分的百分比

        public Split(int d, int p) {
            this.diskIdx = d;
            this.portion = p;
        }

        @Override
        public String toString() {
            return "Split 磁盘ID：" + diskIdx + ", 百分比: " + portion + "%";
        }
    }

    /**
     * Main调用这个方法，传入tagValues
     * 
     * @param tagValues
     * @return
     */
    public static Map<Integer, List<Split>> entrypoint(int[] tagValues) {
        // Set the tags array with the input values
        tags = tagValues;

        Map.Entry<Map<Integer, List<Split>>, Double> result = findOptimalDistributionMT();
        Map<Integer, List<Split>> bestDistribution = result.getKey();

        return bestDistribution;
    }

    /**
     * 本地测试用，默认使用sample_practice.txt处理后拿到的数据
     * 
     * @param args
     */
    public static void main(String[] args) {
        System.out.println("使用 " + Runtime.getRuntime().availableProcessors() + " 个线程进行计算");
        long startTime = System.currentTimeMillis();

        Map.Entry<Map<Integer, List<Split>>, Double> result = findOptimalDistributionMT();
        Map<Integer, List<Split>> bestDistribution = result.getKey();
        double bestVariance = result.getValue();

        long endTime = System.currentTimeMillis();
        printResult(bestDistribution);

        System.out.println("\n程序运行时间: " + (endTime - startTime) + " 毫秒");
    }

    // 计算磁盘负载
    public static List<Double> calculateDiskLoads(Map<Integer, List<Split>> distribution) {
        List<Double> diskLoads = new ArrayList<>(Collections.nCopies(10, 0.0));
        for (Map.Entry<Integer, List<Split>> entry : distribution.entrySet()) {
            double tagSize = tags[entry.getKey()];
            for (Split s : entry.getValue()) {
                diskLoads.set(s.diskIdx, diskLoads.get(s.diskIdx) + tagSize * s.portion / 100.0);
            }
        }
        return diskLoads;
    }

    // 检查分配是否有效
    public static boolean isValidDistribution(Map<Integer, List<Split>> distribution) {
        int[] diskTags = new int[10];
        Arrays.fill(diskTags, 0);

        for (Map.Entry<Integer, List<Split>> entry : distribution.entrySet()) {
            int totalPortion = 0;
            for (Split s : entry.getValue()) {
                if (s.diskIdx >= 10) {
                    return false;
                }
                diskTags[s.diskIdx]++;
                totalPortion += s.portion;
            }
            if (totalPortion != 100) {
                return false;
            }
        }

        for (int count : diskTags) {
            if (count < 1 || count > 4) {
                return false;
            }
        }
        return true;
    }

    // 计算评分（考虑多个因素）
    public static double calculateScore(List<Double> diskLoads) {
        double totalLoad = 0.0;
        double minLoad = Double.POSITIVE_INFINITY;
        double maxLoad = 0.0;

        for (double load : diskLoads) {
            totalLoad += load;
            if (load < minLoad)
                minLoad = load;
            if (load > maxLoad)
                maxLoad = load;
        }

        double mean = totalLoad / 10.0;
        double variance = 0.0;
        double maxDeviation = 0.0;

        for (double load : diskLoads) {
            double diff = load - mean;
            variance += diff * diff;
            double absDiff = Math.abs(diff);
            if (absDiff > maxDeviation)
                maxDeviation = absDiff;
        }

        // 调整权重以更注重均衡性
        return variance / 10.0 + maxDeviation * 3.0 + (maxLoad - minLoad) * 2.0;
    }

    // 生成所有可能的分割方案
    public static List<List<Integer>> generateAllPortions(int numSplits) {
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();

        class Helper {
            void generateRecursive(List<Integer> curr, int remaining, int partsLeft, int minVal) {
                if (partsLeft == 1) {
                    if (remaining >= minVal) {
                        curr.add(remaining);
                        result.add(new ArrayList<>(curr));
                        curr.remove(curr.size() - 1);
                    }
                    return;
                }
                for (int i = minVal; i <= remaining - minVal * (partsLeft - 1); i++) {
                    curr.add(i);
                    generateRecursive(curr, remaining - i, partsLeft - 1, minVal);
                    curr.remove(curr.size() - 1);
                }
            }
        }

        new Helper().generateRecursive(current, 100, numSplits, 1);
        return result;
    }

    // 模拟退火算法的温度函数
    public static double temperature(int iteration, int maxIterations) {
        return Math.pow(0.99, iteration * (double) maxIterations / 1000000.0);
    }

    // 初始化一个基本的分配方案
    public static Map<Integer, List<Split>> initializeDistribution() {
        Map<Integer, List<Split>> dist = new HashMap<>();
        List<Double> diskLoads = new ArrayList<>(Collections.nCopies(10, 0.0));
        int[] diskTagCounts = new int[10];
        Arrays.fill(diskTagCounts, 0);

        // 按大小排序 tags
        List<int[]> sortedTags = new ArrayList<>();
        for (int i = 0; i < tags.length; i++) {
            sortedTags.add(new int[] { tags[i], i });
        }
        // 降序排序
        sortedTags.sort((a, b) -> Integer.compare(b[0], a[0]));

        // 贪心分配
        for (int[] pair : sortedTags) {
            int size = pair[0];
            int tagIdx = pair[1];
            int bestDisk = -1;
            double minScore = Double.POSITIVE_INFINITY;

            for (int disk = 0; disk < 10; disk++) {
                if (diskTagCounts[disk] >= 4)
                    continue;

                double oldVal = diskLoads.get(disk);
                diskLoads.set(disk, oldVal + size);
                double score = calculateScore(diskLoads);
                diskLoads.set(disk, oldVal);

                if (score < minScore) {
                    minScore = score;
                    bestDisk = disk;
                }
            }
            List<Split> splits = new ArrayList<>();
            splits.add(new Split(bestDisk, 100));
            dist.put(tagIdx, splits);

            diskLoads.set(bestDisk, diskLoads.get(bestDisk) + size);
            diskTagCounts[bestDisk]++;
        }
        return dist;
    }

    // 尝试优化分配方案
    public static void optimizeDistribution(int threadId, int numThreads) {
        Random gen = new Random();
        // 预生成所有可能的分割方案
        List<List<List<Integer>>> allPortions = new ArrayList<>();
        for (int i = 1; i <= 4; i++) {
            allPortions.add(generateAllPortions(i));
        }

        Map<Integer, List<Split>> currentDist = initializeDistribution();
        double localBestScore = calculateScore(calculateDiskLoads(currentDist));
        Map<Integer, List<Split>> localBestDist = copyDistribution(currentDist);

        int stagnationCount = 0;
        int iterationsPerThread = 100000;
        int tagsSize = tags.length;

        for (int iter = 0; iter < iterationsPerThread; iter++) {
            double temp = temperature(iter, iterationsPerThread);
            Map<Integer, List<Split>> newDist = copyDistribution(currentDist);

            int numTagsToOptimize = (iter % 4) + 1;
            Set<Integer> selectedTags = new HashSet<>();
            for (int i = 0; i < numTagsToOptimize; i++) {
                int tagIdx;
                do {
                    tagIdx = gen.nextInt(tagsSize);
                } while (selectedTags.contains(tagIdx));
                selectedTags.add(tagIdx);
            }

            for (int tagIdx : selectedTags) {
                List<Split> splits = newDist.get(tagIdx);
                // 随机选择分割数量
                int numSplits = gen.nextInt(4) + 1;
                List<List<Integer>> portions = allPortions.get(numSplits - 1);
                if (portions.isEmpty())
                    continue;

                List<Integer> selectedPortions = portions.get(gen.nextInt(portions.size()));
                splits.clear();

                List<Integer> availableDisks = new ArrayList<>();
                for (int i = 0; i < 10; i++) {
                    availableDisks.add(i);
                }
                Collections.shuffle(availableDisks, gen);

                for (int i = 0; i < numSplits; i++) {
                    splits.add(new Split(availableDisks.get(i), selectedPortions.get(i)));
                }
            }

            if (isValidDistribution(newDist)) {
                double newScore = calculateScore(calculateDiskLoads(newDist));
                double currentScore = calculateScore(calculateDiskLoads(currentDist));
                double delta = newScore - currentScore;

                if (delta < 0 || (temp > 0 && Math.exp(-delta / temp) > gen.nextDouble())) {
                    currentDist = newDist;
                    currentScore = newScore;

                    if (newScore < localBestScore) {
                        localBestScore = newScore;
                        localBestDist = copyDistribution(newDist);
                        stagnationCount = 0;
                    }

                    mtx.lock();
                    try {
                        if (newScore < globalBestVariance) {
                            globalBestVariance = newScore;
                            copyIntoGlobalBest(newDist);
                        }
                    } finally {
                        mtx.unlock();
                    }
                }
            }

            stagnationCount++;
            if (stagnationCount > 10000) {
                currentDist = initializeDistribution();
                stagnationCount = 0;
            }
        }
    }

    // 多线程优化
    public static Map.Entry<Map<Integer, List<Split>>, Double> findOptimalDistributionMT() {
        int numThreads = Runtime.getRuntime().availableProcessors();
        List<Thread> threads = new ArrayList<>();

        Map<Integer, List<Split>> initDist = initializeDistribution();
        globalBestVariance = calculateScore(calculateDiskLoads(initDist));
        copyIntoGlobalBest(initDist);

        for (int i = 0; i < numThreads; i++) {
            final int threadId = i;
            Thread t = new Thread(() -> optimizeDistribution(threadId, numThreads));
            threads.add(t);
            t.start();
        }

        for (Thread t : threads) {
            try {
                t.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }

        return new AbstractMap.SimpleEntry<>(copyDistribution(globalBestDistribution),
                globalBestVariance);
    }

    // 打印结果
    private static void printResult(Map<Integer, List<Split>> distribution) {
        if (distribution.isEmpty()) {
            System.out.println("未找到有效解");
            return;
        }

        List<Double> diskLoads = calculateDiskLoads(distribution);
        List<List<int[]>> diskTags = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            diskTags.add(new ArrayList<>());
        }

        for (Map.Entry<Integer, List<Split>> entry : distribution.entrySet()) {
            int tagIdx = entry.getKey();
            for (Split s : entry.getValue()) {
                double portion = s.portion / 100.0;
                diskTags.get(s.diskIdx).add(new int[] { tagIdx, (int) (portion * tags[tagIdx]) });
            }
        }

        double totalLoad = 0.0;
        double minLoad = Double.POSITIVE_INFINITY;
        double maxLoad = 0.0;
        for (double load : diskLoads) {
            totalLoad += load;
            if (load < minLoad)
                minLoad = load;
            if (load > maxLoad)
                maxLoad = load;
        }

        double meanLoad = totalLoad / 10.0;
        double variance = 0.0;
        for (double load : diskLoads) {
            double diff = load - meanLoad;
            variance += diff * diff;
        }
        variance /= 10.0;

        DecimalFormat df = new DecimalFormat("0.00");
        System.out.println("\n负载统计：");
        System.out.println("平均负载: " + df.format(meanLoad));
        System.out.println("最小负载: " + minLoad);
        System.out.println("最大负载: " + maxLoad);
        System.out.println("负载差异: " + (maxLoad - minLoad));
        System.out.println("方差: " + variance);
        System.out.println("优化目标值: " + globalBestVariance);

        System.out.println("\n最优分配方案：");
        for (int diskIdx = 0; diskIdx < 10; diskIdx++) {
            System.out.println("\n磁盘 " + diskIdx + ":");
            System.out.println("总负载: " + df.format(diskLoads.get(diskIdx)));
            System.out.println("Tags:");
            for (int[] info : diskTags.get(diskIdx)) {
                int tagIdx = info[0];
                double portionLoad = info[1];
                double originalTag = tags[tagIdx];
                double fraction = portionLoad / originalTag;
                System.out.println("  Tag " + tagIdx + ": " + df.format(fraction) + " * "
                        + (int) originalTag + " = " + df.format(portionLoad));
            }
        }
    }

    private static Map<Integer, List<Split>> copyDistribution(Map<Integer, List<Split>> source) {
        Map<Integer, List<Split>> copy = new HashMap<>();
        for (Map.Entry<Integer, List<Split>> e : source.entrySet()) {
            List<Split> newSplits = new ArrayList<>();
            for (Split s : e.getValue()) {
                newSplits.add(new Split(s.diskIdx, s.portion));
            }
            copy.put(e.getKey(), newSplits);
        }
        return copy;
    }

    private static void copyIntoGlobalBest(Map<Integer, List<Split>> source) {
        globalBestDistribution.clear();
        for (Map.Entry<Integer, List<Split>> e : source.entrySet()) {
            List<Split> newSplits = new ArrayList<>();
            for (Split s : e.getValue()) {
                newSplits.add(new Split(s.diskIdx, s.portion));
            }
            globalBestDistribution.put(e.getKey(), newSplits);
        }
    }
}
